oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/JwtEncoderParameters.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2002-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.jwt;
+
+import org.springframework.util.Assert;
+
+/**
+ * A holder of parameters containing the JOSE header and JWT Claims Set.
+ *
+ * @author Joe Grandja
+ * @since 5.6
+ * @see JwtEncoder
+ */
+public final class JwtEncoderParameters {
+
+	private final JoseHeader headers;
+
+	private final JwtClaimsSet claims;
+
+	private JwtEncoderParameters(JoseHeader headers, JwtClaimsSet claims) {
+		Assert.notNull(headers, "headers cannot be null");
+		Assert.notNull(claims, "claims cannot be null");
+		this.headers = headers;
+		this.claims = claims;
+	}
+
+	/**
+	 * Returns a new {@link JwtEncoderParameters}, initialized with the provided
+	 * {@link JoseHeader} and {@link JwtClaimsSet}.
+	 * @param headers the {@link JoseHeader}
+	 * @param claims the {@link JwtClaimsSet}
+	 * @return the {@link JwtEncoderParameters}
+	 */
+	public static JwtEncoderParameters with(JoseHeader headers, JwtClaimsSet claims) {
+		return new JwtEncoderParameters(headers, claims);
+	}
+
+	/**
+	 * Returns the {@link JoseHeader headers}.
+	 * @return the {@link JoseHeader}
+	 */
+	public JoseHeader getHeaders() {
```

Comment:
Thanks for looking into this. I think my motivations here are tied to wanting a contract that can support Nested JWTs, so I will focus my feedback along those lines.

To be clear, I am imagining that `JwtEncoderParameters` could eventually look like the following (assuming renaming of classes):

```java
JwtClaimsSet getClaims();
JwsHeader getJwsHeader();
JweHeader getJweHeader();
```

This, of course, makes sense in the context of Nested JWTs -- for the contract to support sign-then-encrypt, it would be important to be able to supply both signing headers and encryption headers in the method call.

> I suppose we could have `getJwsHeader()` and `getJweHeader()` but that would not be a clean design from my perspective.

Can you elaborate on why this isn't a clean design from your perspective? I may be able to help if I can understand the specific issue.

Reply:
> Can you elaborate on why this isn't a clean design from your perspective?

A JWT is either a JWS or a JWE, not both. So having `getJwsHeader()` and `getJweHeader()` does not make sense from a design standpoint. Even in the Nested JWT use case (sign then encrypt), it's still a JWE but the payload is a JWS.

Is the user to assume that `getJwsHeader()` is dual purpose? The header for a JWS OR the header for a nested JWS?

Reply:
> A JWT is either a JWS or a JWE, not both

I agree with this, though I'm not seeing the connection to `JwtEncoderParameters`. This class is not trying to represent a JWT but instead the parameters needed to encode a JWT. It may even contain methods and fields that are not headers or claims.

> Is the user to assume that `getJwsHeader()` is dual purpose? 

The user need not assume anything since they are using factory methods to describe their intent.

The implementation then reads the parameters object to infer the user's intent. If the parameters object indicates that signing should be done, then the implementation does so. After that, if the parameters object also indicates that encryption should be done, then the implementation does that, too.

Reply:
I think we could possibly be going afield from [the main reason](https://github.com/spring-projects/spring-security/pull/9208#discussion_r710349600) (see the second-to-last paragraph) for my suggestion, though. We can certainly make anything work - we want to make sure that we are building the right thing.

IMO, `JwtDecoder` and `JwtEncoder` should have comparable capabilities. [`JwtDecoder`'s contract supports a Nested JWT](https://github.com/spring-projects/spring-security-samples/blob/main/servlet/spring-boot/java/oauth2/resource-server/jwe/src/main/java/example/OAuth2ResourceServerSecurityConfiguration.java#L88) without additional interfaces or method calls, and so it seems that `JwtEncoder` should be able to as well.